celery/utils/functional.py

```diff
@@ -189,38 +205,73 @@ def __init__(self, it):
         self.__it = it
         self.__index = 0
         self.__consumed = []
+        self.__done = False
 
     def __reduce__(self):
         return list, (self.data,)
 
     def __length_hint__(self):
         return self.__it.__length_hint__()
 
+    def __repr__(self):
+        # override list.__repr__ to avoid consuming the generator
+        if self.__done:
+            return repr(self.__consumed)
+        else:
+            return '[{0}]'.format(', '.join(
+                [repr(x) for x in self.__consumed] + ['...'])
+            )
+
     def __iter__(self):
-        return chain(self.__consumed, self.__it)
+        for x in self.__consumed:
+            yield x
+        if not self.__done:
+            for y in self.__it:
+                self.__consumed.append(y)
+                yield y
+            self.__done = True
 
     def __getitem__(self, index):
         if index < 0:
             return self.data[index]
         try:
             return self.__consumed[index]
         except IndexError:
+            it = iter(self)
             try:
                 for _ in range(self.__index, index + 1):
-                    self.__consumed.append(next(self.__it))
+                    next(it)
             except StopIteration:
                 raise IndexError(index)
             else:
                 return self.__consumed[index]
 
-    @property
-    def data(self):
+    def __nonzero__(self):
+        # nonzero for list calls len() which would consume the generator:
+        # override to consume maximum of one item.
+        if len(self.__consumed):
+            return True
         try:
-            self.__consumed.extend(list(self.__it))
+            next(iter(self))
```

Comment:
It's not clear that an item is not lost because of https://github.com/celery/celery/pull/4459/files#diff-427d1dbcd92e9e366527f44ff54821a1R226
Please add a comment explaining what happens when you do so.